angel-ps/mllib/src/main/java/com/tencent/angel/ml/warplda/Trainer.scala

```diff
@@ -0,0 +1,500 @@
+
+    while (futures.nonEmpty) {
+      val keys = futures.keySet.iterator
+      while (keys.hasNext) {
+        val pkey = keys.next()
+        val future = futures(pkey)
+        if (future.isDone) {
+          val sampler = queue.take()
+          future.get() match {
+            case csr: PartCSRResult => executor.execute(new Task(sampler, pkey, csr))
+            case _ => throw new AngelException("should by PartCSRResult")
+          }
+          futures.remove(pkey)
+        }
+      }
+    }
+
+    for (i <- 0 until model.threadNum) queue.take()
+
+    model.tMat.clock(false).get()
+  }
+
+
+
+  def scheduleReset(): Unit = {
+    class Task(sampler: Sampler, pkey: PartitionKey) extends Thread {
+      override def run(): Unit = {
+        sampler.reset(pkey)
+        queue.add(sampler)
+      }
+    }
+
+    util.Arrays.fill(nk, 0)
+    for (i <- 0 until model.threadNum) queue.add(new Sampler(data, model).set(nk))
+    val iter = pkeys.iterator()
+    while (iter.hasNext) {
+      val sampler = queue.take()
+      executor.execute(new Task(sampler, iter.next()))
+    }
+
+    // calculate the delta value of nk
+    // the take means that all tasks have been finished
+    val update = new DenseIntVector(model.K)
+    for (i <- 0 until model.threadNum) {
+      val sampler = queue.take()
+      for (i <- 0 until model.K)
+        update.plusBy(i, sampler.nk(i) - nk(i))
+    }
+
+    model.tMat.increment(0, update)
+    // update for wt
+    model.wtMat.clock().get()
+    // update for nk
+    model.tMat.clock().get()
+  }
+
+
+
+  def computeWordLLH: Double = {
+    model.wtMat.get(new LikelihoodFunc(model.wtMat.getMatrixId(), beta)) match {
+      case r : ScalarAggrResult => r.getResult
+      case _ => throw new AngelException("should be ScalarAggrResult")
+    }
+  }
+
+  def computeWordLLHSummary: Double = {
+    var ll = model.K * Gamma.logGamma(beta * model.V)
+    for (k <- 0 until model.K)
+      ll -= Gamma.logGamma(nk(k) + beta * model.V)
+    ll
+  }
+
+  def likelihood: Double = {
+    var ll = 0d
+    fetchNk
+    if (ctx.getTaskIndex == 0)
+      ll += computeWordLLHSummary + computeWordLLH
+    ll += scheduleDocllh(data.n_docs)
+    ll
+  }
+
+  def scheduleDocllh(n_docs: Int):Double = {
+    val results = new LinkedBlockingQueue[Double]()
+    val nnzs = new LinkedBlockingQueue[Int]()
+    class Task(index: AtomicInteger) extends Thread {
+      private var ll = 0.0
+      private var nnz = 0
+      override def run(): Unit = {
+        while (index.get() < n_docs) {
+          val d = index.incrementAndGet()
+          val dk = mutable.Map[Int,Int]()
+          if (d < n_docs) {
+            (data.accDoc(d) until data.accDoc(d+1)) foreach {j=>
+              val k = data.topics(data.inverseMatrix(j))
+              dk += k ->(dk.getOrElse(k, 0) + 1)
+            }
+            dk.foreach{case(k, value) =>
+              ll += Gamma.logGamma(alpha + dk(value))
```

Comment:
this should be  "ll += Gamma.logGamma(alpha + value)"

Reply:
Yeah, it's really a typo. Are your initialization get stucked as well? I'm even hanging during the during the initialization